modmail/extensions/utils/error_handler.py

```diff
@@ -0,0 +1,195 @@
+import logging
+import re
+import typing
+
+import discord
+import discord.errors
+from discord.ext import commands
+
+from modmail.bot import ModmailBot
+from modmail.log import ModmailLogger
+from modmail.utils.cogs import BotModes, ExtMetadata, ModmailCog
+from modmail.utils.extensions import BOT_MODE
+
+
+logger: ModmailLogger = logging.getLogger(__name__)
+
+EXT_METADATA = ExtMetadata()
+
+ERROR_COLOUR = discord.Colour.red()
+
+ERROR_TITLE_REGEX = re.compile(r"(?<=[a-zA-Z])([A-Z])(?=[a-z])")
+
+ANY_DEV_MODE = BOT_MODE & (BotModes.DEVELOP.value + BotModes.PLUGIN_DEV.value)
+
+
+class ErrorHandler(ModmailCog, name="Error Handler"):
+    """Handles all errors across the bot."""
+
+    def __init__(self, bot: ModmailBot):
+        self.bot = bot
+
+    @staticmethod
+    def error_embed(message: str, title: str = None) -> discord.Embed:
+        """Create an error embed with an error colour and reason and return it."""
+        return discord.Embed(message, colour=ERROR_COLOUR, title=title or "Error Occurred")
+
+    @staticmethod
+    def get_title_from_name(error: typing.Union[Exception, str]) -> str:
+        """
+        Return a message dervived from the exception class name.
+
+        Eg NSFWChannelRequired returns NSFW Channel Required
+        """
+        if not isinstance(error, str):
+            error = error.__class__.__name__
+        return re.sub(ERROR_TITLE_REGEX, r" \1", error)
+
+    @staticmethod
+    def _reset_command_cooldown(ctx: commands.Context) -> bool:
+        if return_value := ctx.command.is_on_cooldown(ctx):
+            ctx.command.reset_cooldown(ctx)
+        return return_value
+
+    async def handle_user_input_error(
+        self,
+        ctx: commands.Context,
+        error: commands.UserInputError,
+        reset_cooldown: bool = True,
+    ) -> discord.Embed:
+        """Handling deferred from main error handler to handle UserInputErrors."""
+        if reset_cooldown:
+            self._reset_command_cooldown(ctx)
+        embed = None
+        msg = None
+        title = "User Input Error"
+        if isinstance(error, commands.BadUnionArgument):
+            msg = self.get_title_from_name(str(error))
+            title = self.get_title_from_name(error)
+        else:
+            title = self.get_title_from_name(error)
+        return embed or self.error_embed(msg or str(error), title=title)
```

Comment:
This can be cleaned up a little. Try moving `str(error)` into the `else` clause, and then you don't need the or. Also no need for `title` here to be a keyword argument.